src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java

```diff
@@ -4975,14 +5031,290 @@ public String toString() {
         }
     }
 
-    //-------------------------------------------------------------------------
+    //-----------------------------------------------------------------------
+
     /**
-     * Length comparator.
+     * Prints or parses day periods.
      */
-    static final Comparator<String> LENGTH_SORT = new Comparator<String>() {
+    static final class DayPeriodPrinterParser implements DateTimePrinterParser {
+        private final TextStyle textStyle;
+        private final static ConcurrentMap<Locale, LocaleStore> DAYPERIOD_LOCALESTORE = new ConcurrentHashMap<>();
+
+        /**
+         * Constructor.
+         *
+         * @param textStyle  the text style, not null
+         */
+        DayPeriodPrinterParser(TextStyle textStyle) {
+            // validated by caller
+            this.textStyle = textStyle;
+        }
+
         @Override
-        public int compare(String str1, String str2) {
-            return str1.length() == str2.length() ? str1.compareTo(str2) : str1.length() - str2.length();
+        public boolean format(DateTimePrintContext context, StringBuilder buf) {
+            Long value = context.getValue(MINUTE_OF_DAY);
```

Comment:
This does not match the spec: " During formatting, the day period is obtained from {@code HOUR_OF_DAY}, and optionally {@code MINUTE_OF_HOUR} if exist"

It is possible and legal to create a Temporal that returns `HOUR_OF_DAY` and `MINUTE_OF_HOUR` but not `MINUTE_OF_DAY`. As such, this method must be changed to follow the spec.

-----

In addition, it is possible for `HOUR_OF_DAY` and `MINUTE_OF_HOUR` to be outside their normal bounds. The right behaviour would be to combine the two fields within this method, and then use mod to get the value into the range 0 to 1440 before calling `dayPeriod.include`. (While the fall back behaviour below does cover this, it would be better to do what I propose here.)

An example of this is a `TransportTime` class where the day runs from 03:00 to 27:00 each day (because trains run after midnight for no extra cost to the passenger, and it is more convenient for the operator to treat the date that way). A `TransportTime` of 26:30 should still resolve to "night1" rather than fall back to "am".

Reply:
Fixed.